h2o-r/h2o-package/R/models.R

```diff
@@ -5613,19 +5613,27 @@ h2o.cross_validation_predictions <- function(object) {
 #' iris_gbm <- h2o.gbm(x = c(1:4), y = 5, training_frame = iris_hex)
 #'
 #' # one target class
-#' h2o.partialPlot(object = iris_gbm, data = iris_hex, cols="Petal.Length", targets=c("setosa"))
+#' h2o.partialPlot(object = iris_gbm, newdata = iris_hex, cols="Petal.Length", targets=c("setosa"))
 #' # three target classes
-#' h2o.partialPlot(object = iris_gbm, data = iris_hex, cols="Petal.Length", 
+#' h2o.partialPlot(object = iris_gbm, newdata = iris_hex, cols="Petal.Length", 
 #'                  targets=c("setosa", "virginica", "versicolor"))
 #' }
 #' @export
 
-h2o.partialPlot <- function(object, data, cols, destination_key, nbins=20, plot = TRUE, plot_stddev = TRUE,
-                            weight_column=-1, include_na=FALSE, user_splits=NULL, col_pairs_2dpdp=NULL, save_to=NULL, 
-                            row_index=-1, targets=NULL) {
+h2o.partialPlot <- function(object, newdata, cols, destination_key, nbins=20, plot = TRUE, plot_stddev = TRUE,
+                            weight_column=-1, include_na=FALSE, user_splits=NULL, col_pairs_2dpdp=NULL, save_to=NULL,
+                            row_index=-1, targets=NULL, ...) {
+  varargs <- list(...)
+  for (arg in names(varargs)) {
+      if (arg == 'data') {
+          warning("argument 'data' is deprecated; please use 'newdata' instead.")
+          if (missing(newdata))
+              newdata <- varargs$data else warning("ignoring 'data' as 'newdata' was also provided.")
+      }
+  }
```

Comment:
This is a bit tricky. R is lazy in evaluating its arguments and using `list(...)` evaluates all the things in the `...`.

For example:
```r
abc <- function (newdata, ...) {
  varargs <- list(...)
  for (arg in names(varargs)) {
    if (arg == 'data') {
      warning("argument 'data' is deprecated; please use 'newdata' instead.")
      if (missing(newdata))
        newdata <- varargs$data else warning("ignoring 'data' as 'newdata' was also provided.")
    }
  }
  newdata
}

abx <- function (newdata, ...) {
  if ('data' %in% ...names()) {
      warning("argument 'data' is deprecated; please use 'newdata' instead.")
      if (missing(newdata))
        newdata <- ...elt(which(...names() == 'data')) else warning("ignoring 'data' as 'newdata' was also provided.")
  }
  newdata
}


abc(1) == abx(1) # TRUE
abc(1, data = 2) == abx(1, data = 2) # TRUE
abc(data = 2) == abx(data = 2) # TRUE
abc(2, lazy_R_says = stop("evaluating unnecessary things")) # Stops 
abx(2, lazy_R_says = stop("evaluating unnecessary things")) # Doesn't evaluate args that it should not care about
abc(data = 2, lazy_R_says = stop("evaluating unnecessary things")) # Stops
abx(data = 2, lazy_R_says = stop("evaluating unnecessary things")) # Doesn't evaluate args that it should not care about
```

You can either use the combination of `...names` and `...elt` or something like `match.call(expand.dots = TRUE)$data` (but the former seems easier to me in this case).

Reply:
great comment, thanks! I was not aware of `...names` or `...elt`, definitely much better here.
I think the `list(...)` pattern has been used in other places for the same goal so I'll fix it there as well.

Reply:
@tomasfryda I fixed this particular occurence, but decided not to fix others in this ticket as there are much more suboptimal usages of `list(...)` than I though in our codebase. The rest is moved to https://github.com/h2oai/h2o-3/issues/15905

btw, funnily the R doc is giving a better advice in its `dots` (..., ..1, etc used in Functions) section:
> ...length() returns the number of expressions in ..., and ...names() the [names](http://127.0.0.1:17473/help/library/base/help/names). These are the same as length(list(...)) or names(list(...)) but without evaluating the expressions in ... (which happens with list(...)).

than in its `dotMethods` (The Use of ... in Method Signatures) section:
> Typically the methods will pass “...” down to another function or will create a list of the arguments and iterate over that. See the examples below.